File: samsung/java/BOJ17822.java
import java.util.*;
import java.io.*;

public class BOJ17822 {
    static int N, M, T, x, k, d;
    static int[][] commands;
    static ArrayList<Integer>[] circles;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] line = br.readLine().split(" ");

        N = Integer.parseInt(line[0]);
        M = Integer.parseInt(line[1]);
        T = Integer.parseInt(line[2]);

        circles = new ArrayList[N+1];
        commands = new int[T][3];


        for(int i=1;i<=N;i++){
            line = br.readLine().split(" ");
            circles[i] = new ArrayList<>();
            for(int j=0;j<M;j++){
                circles[i].add(Integer.parseInt(line[j]));
            }
        }

        for(int i=0;i<T;i++){
            line = br.readLine().split(" ");
            for(int j=0;j<3;j++){
                commands[i][j] = Integer.parseInt(line[j]);
            }
        }

        // for(ArrayList<Integer> circle : circles){
        //     System.out.println(circle);
        // }
        // System.out.println();
        solve();
    }
    public static void solve(){
        for(int[] command : commands){
            rotate(command[0], command[1], command[2]);
            delete();   
            // for(ArrayList<Integer> circle : circles){
            //     System.out.println(circle);
            // }
            // System.out.println();
        }
        System.out.println(getSum());
    }
    public static int getSum(){
        int ans = 0;
        for(int i=1;i<=N;i++){
            for(int j=0;j<M;j++){
                ans += circles[i].get(j);
            }
        }
        return ans;
    }
    public static void delete(){
        double sum = 0;
        int cnt = 0;
        boolean[][] flags = new boolean[N+1][M];
        boolean isDeleted = false;
        for(int i=1;i<=N;i++){
            if(i % 2 == 0){
                int up = i + 1;
                int down = i - 1;
                for(int j=0;j<M;j++){
                    int cur = circles[i].get(j);
                    if(cur == 0) continue;
                    if(up <= N && cur == circles[up].get(j)){
                        flags[up][j] = true;
                        flags[i][j] = true;
                        isDeleted = true;
                    }
                    if(cur == circles[down].get(j)){
                        flags[down][j] = true;
                        flags[i][j] = true;
                        isDeleted = true;
                    }
                }
            }

            for(int j=0;j<M;j++){
                int cur = circles[i].get(j);
                if (cur == 0) continue;
                sum += circles[i].get(j);
                cnt ++;
                if(cur == circles[i].get((j+1) % M)){
                    flags[i][j] = true;
                    flags[i][(j+1) % M] = true;
                    isDeleted = true;
                }
            }
        }

        if(isDeleted){
            for(int i=1;i<=N;i++){
                for(int j=0;j<M;j++){
                    if(flags[i][j]){
                        circles[i].set(j, 0);
                    }
                }
            }
        } else {
            if(cnt != 0){
                double mean = sum / cnt;
                for(int i=1;i<=N;i++){
                    for(int j=0;j<M;j++){
                        int cur = circles[i].get(j);
                        if(cur == 0) continue;
                        if(cur < mean){
                            circles[i].set(j, cur + 1);
                        } else if(cur > mean){
                            circles[i].set(j, cur - 1);
                        }
                    }
                }   
            }
        }
    }
    public static void rotate(int x, int d, int k){
        for(int i=x;i<=N;i+=x){
            if(d == 0){
                for(int j=0;j<k;j++){
                    circles[i].add(0, circles[i].remove(M-1));
                }
            } else {
                for(int j=0;j<k;j++){
                    circles[i].add(M-1, circles[i].remove(0));
                }
            }
        }
    }
}

File: samsung/BOJ17822.py
from sys import stdin
from collections import deque
readline = stdin.readline

N, M, T = map(int, readline().split())
circles = [deque() for _ in range(N+1)]

for r in range(1, N+1):
    circle = circles[r]
    line = list(map(int, readline().split()))
    for l in line:
        circle.append(l)

commands = [list(map(int, readline().split())) for _ in range(T)]

def solve():
    for command in commands:
        rotate(*command)
        delete()


    total = 0
    for i in range(1, N+1):
        total += sum(circles[i])
    print(total)

def delete():
    # 상하 인접
    # 짝수 원판의 아래 위만 검사하여 제거
    flags = [[False] * M for _ in range(N+1)]
    isDeleted = False
    total = 0
    cnt = 0
    for i in range(1, N+1):
        if i % 2 == 0:
            up = i + 1
            down = i - 1
            
            for j in range(M):
                if circles[i][j] == 0: continue
                if up <= N and circles[i][j] == circles[up][j]:
                    flags[up][j] = True
                    flags[i][j] = True
                if circles[i][j] == circles[down][j]:
                    flags[i][j] = True
                    flags[down][j] = True
        
                if flags[i][j]:
                    isDeleted = True

        for j in range(M):
            if circles[i][j] == 0: continue
            total += circles[i][j]
            cnt += 1
            if circles[i][j] == circles[i][(j+1) % M]:
                flags[i][j] = True
                flags[i][(j+1) % M] = True

            if flags[i][j]:
                isDeleted = True

    if isDeleted:
        for r in range(1, N+1):
            for c in range(M):
                if flags[r][c]:
                    circles[r][c] = 0
    else:
        if cnt:
            mean = total / cnt
            for r in range(1, N+1):
                for c in range(M):
                    if circles[r][c]:
                        if circles[r][c] < mean:
                            circles[r][c] += 1
                        elif circles[r][c] > mean:
                            circles[r][c] -= 1
def rotate(x, d, k):
    for i in range(1, len(circles)):
        if i % x == 0:
            # 시계 방향
            if d == 0:
                for _ in range(k):
                    circles[i].appendleft(circles[i].pop())
            else:
                for _ in range(k):
                    circles[i].append(circles[i].popleft())

solve()